packages/datadog-instrumentations/src/fs.js

```diff
@@ -0,0 +1,357 @@
+  utimes: ['path', 'atime', 'mtime'],
+  write: ['fd'],
+  writeFile: ['file', 'data', 'options'],
+  writev: ['fd']
+}
+
+const watchMethods = {
+  unwatchFile: ['path', 'listener'],
+  watch: ['path', 'options', 'listener'],
+  watchFile: ['path', 'options', 'listener']
+}
+
+const paramsByFileHandleMethods = {
+  appendFile: ['data', 'options'],
+  chmod: ['mode'],
+  chown: ['uid', 'gid'],
+  close: [],
+  createReadStream: ['options'],
+  createWriteStream: ['options'],
+  datasync: [],
+  read: ['buffer', 'offset', 'length', 'position'],
+  readableWebStream: [],
+  readFile: ['options'],
+  readLines: ['options'],
+  readv: ['buffers', 'position'],
+  stat: ['options'],
+  sync: [],
+  truncate: ['len'],
+  utimes: ['atime', 'mtime'],
+  write: ['buffer', 'offset', 'length', 'position'],
+  writeFile: ['data', 'options'],
+  writev: ['buffers', 'position']
+}
+
+addHook({ name: 'fs' }, fs => {
+  const asyncMethods = Object.keys(paramsByMethod)
+  const syncMethods = asyncMethods.map(name => `${name}Sync`)
+
+  massWrap(fs, asyncMethods, createWrapFunction())
+  massWrap(fs, syncMethods, createWrapFunction())
+  massWrap(fs.promises, asyncMethods, createWrapFunction('promises.'))
+
+  wrap(fs.realpath, 'native', createWrapFunction('', 'realpath.native'))
+  wrap(fs.realpathSync, 'native', createWrapFunction('', 'realpath.native'))
+  wrap(fs.promises.realpath, 'native', createWrapFunction('', 'realpath.native'))
+
+  wrap(fs, 'createReadStream', createWrapCreateStream())
+  wrap(fs, 'createWriteStream', createWrapCreateStream())
+  if (fs.Dir) {
+    wrap(fs.Dir.prototype, 'close', createWrapFunction('dir.'))
+    wrap(fs.Dir.prototype, 'closeSync', createWrapFunction('dir.'))
+    wrap(fs.Dir.prototype, 'read', createWrapFunction('dir.'))
+    wrap(fs.Dir.prototype, 'readSync', createWrapFunction('dir.'))
+    wrap(fs.Dir.prototype, Symbol.asyncIterator, createWrapDirAsyncIterator())
+  }
+
+  wrap(fs, 'unwatchFile', creaetWatchWrapFunction())
+  wrap(fs, 'watch', creaetWatchWrapFunction())
+  wrap(fs, 'watchFile', creaetWatchWrapFunction())
+
+  wrapFileHandle(fs)
+  return fs
+})
+
+async function wrapFileHandle (fs) {
+  const fileHandlePrototype = await getFileHandlePrototype(fs)
+  const desc = Reflect.getOwnPropertyDescriptor(fileHandlePrototype, kHandle)
+  if (!desc || !desc.get) {
+    Reflect.defineProperty(fileHandlePrototype, kHandle, {
+      get () {
+        return this[ddFhSym]
+      },
+      set (h) {
+        this[ddFhSym] = h
+        wrap(this, 'close', createWrapFunction('filehandle.'))
+      },
+      configurable: true
+    })
+  }
+  for (const name of Reflect.ownKeys(fileHandlePrototype)) {
+    if (typeof name !== 'string' || name === 'constructor' || name === 'fd' || name === 'getAsyncId') {
+      continue
+    }
+    wrap(fileHandlePrototype, name, createWrapFunction('filehandle.'))
+  }
+}
+
+async function getFileHandlePrototype (fs) {
+  const fh = await fs.promises.open(__filename, 'r')
```

Comment:
Patching occurs once at load time, so this might be better done sync. 🤔

Reply:
The purpose of this method is get the prototype of `FileHandle`, this class can be initialized externally, so we need an instance. This class is only returned in promise API of fs:
https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#promises-api

Reply:
In that case, I would just patch it the first time a call to `fsPromise.open(...)` resolves. The way it is now, it's a race condition waiting to happen and could crash the process without a `.catch(...)` on the outer call.

Reply:
Good idea, changed.